Support alarm notifications

dpino opened the pull request:
Alarms spec determines that certain events can generate notifications, for instance, when a new alarm inventory entry is added, when a new alarm is raised or cleared or when a status operator state changes.

This PR adds support for notifications and makes use of it in the alarms module.

The way is working is by adding a pool of generated notifications in alarms' state. The leader pulls the alarm pool and pushes all generated notifications in an interval to the listening peers.

Example of use:

lwAFTR that generates ARP error:

```bash
$ sudo ./snabb lwaftr run --name lwaftr --conf lwaftr-mini.conf --v4 82:00.0 --v6 82:00.1
lwaftr-mini.conf: loading compiled configuration from lwaftr-mini.o                     
lwaftr-mini.conf: compiled configuration is up to date.                                 
Migrating instance 'test' to '82:00.0'      
Warning: No assignable CPUs declared; leaving data-plane process without assigned CPU.  
No CPUs available; not binding to any NUMA node.                                        
ARP: Resolving '10.0.1.2'                   
```

Open a listening peer to instance:

```bash
$ sudo ./snabb config listen lwaftr
```

Change status operator state
```bash
$ sudo ./snabb alarms set-operator-state --schema snabb-softwire-v2 \
   lwaftr 14239 arp-resolution ack
```

Once the operator-state changes to `ack` the listening peer received an _operator_action notification.

```bash
$ sudo ./snabb config listen lwaftr
operator_action_notification
```

dpino committed:
Implement alarm-notification logic

dpino committed:
Basic notification push mechanism

dpino committed:
Fetch notifications from alarms state

dpino commented on src/lib/ptree/ptree.lua:
```diff
@@ -527,6 +527,24 @@ end
+   local notifications = alarms.notifications()
+   while i <= #peers do
+      local peer = peers[i]
+      for _, each in ipairs(notifications) do
```
Somehow here is needed to decode a notifications into a string (representing a JSON object for the corresponding notification), I think. The object should be decoded according to the YANG schema.

dpino committed:
Serialize notifications to JSON

dpino commented:
Added an extra commit that serializes Lua objects (notifications) to JSON.

Setup:

```
$ sudo ./snabb lwaftr run --name lwaftr --conf lwaftr-mini.conf --v4 82:00.0 --v6 82:00.1
ARP: Resolving '10.0.1.2'
```

```bash
$ sudo ./snabb config listen lwaftr
```

```bash
$ sudo ./snabb alarms set-operator-state --schema snabb-softwire-v2 lwaftr 8075 arp-resolution ack
```

Response:

```bash
$ sudo ./snabb config listen lwaftr
{"state":"ack","time":"2018-05-02T16:14:18Z","operator":"admin","text":""}
```

I think with this change, the PR is ready.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -527,6 +528,27 @@ end
 
+local function send_message(socket, msg_str)
+   socket:write(tostring(#msg_str)..'\n'..msg_str)
+end
```
I don't think this is going to work.  The manager supports non-blocking message sends and receives to and from peers.  A peer goes through a number of states: waiting for a call message, reading the message, handling the message, writing the response, and then back to the beginning.  The reading and writing phases are non-blocking; they might complete over a number of turns.

So, notifications would have to fit into this state machine.  You can't just write over a socket; you might be corrupting a concurrent write.

What do you think about adding a separate "snabb alarms listen" command that listens for alarm notifications?  Since it wouldn't be doing anything else, we could just use the existing mechanism to enqueue messages for the peer.

I think the medium term we should switch to use fibers in the manager.  That way when a new peer connects, you can just call `new_peer(socket)`:

```
local function service_peer(socket)
   while true do
      local call = read_call(socket)
      local response = handle_call(call)
      write_response(socket, response)
   end
end

local function new_peer(socket)
  fiber.spawn(service_peer, socket)
end
```

i.e. provide control flow for a synchronous call-handle-reply loop, while the fibers system arranges to suspend fibers if the socket operation would block and resume when the operation can continue.

wingo replied:
I think this still won't work as the socket is marked non-blocking.  You will need to enqueue the message and send it incrementally I think.

dpino committed:
Add alarms listen subprogram

Used to listen to alarm notifications sent by the leader.  A client calls
attach-notification-listener to hook to a leader.  The leader only sends alarm
notifications to those peers that connected via this command.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -491,6 +493,29 @@ function Manager:rpc_attach_listener (args)
+
+function Manager:rpc_attach_notification_listener (args)
+   local function attacher()
+      if self.listen_peer ~= nil then error('Listener already attached') end
```
I think it should be possible to have more than one notification listener; wdyt?

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -491,6 +493,29 @@ function Manager:rpc_attach_listener (args)
+function Manager:rpc_attach_notification_listener (args)
+   local function attacher()
+      if self.listen_peer ~= nil then error('Listener already attached') end
+      self.listen_peer = self.rpc_peer
```
Here there's probably no need to set listen_peer or have any relationship to normal "snabb config listen" code since you add the peer to the notification set.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -527,6 +552,27 @@ end
+   socket:write(tostring(#msg_str)..'\n'..msg_str)
+end
+
+function Manager:push_notifications_to_peers()
```
Each peer will need a queue of outgoing messages and a cursor to mark how many bytes have been written so far, I think; like the other code.  Probably should unify with the other code.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -491,6 +493,27 @@ function Manager:rpc_attach_listener (args)
+      self:add_notification_peer(self.rpc_peer)
+      return {}
+   end
+   local success, response = pcall(attacher)
```
As far as I can tell the "attacher" call can't fail so the pcall isn't needed, and probably add_notification_peer can be inlined here.  I guess the intention is that once you attach for notifications, you no longer can send any RPCs, right?

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -527,8 +550,35 @@ end
+   self:handle_calls_from_peers(self.notification_peers)
+end
+
+function Manager:handle_calls_from_peers(peers)
```
Sorry for being so picky here.  Related to my last comment, I think probably once a peer is listening for notifications, it shouldn't be able to make any more RPCs.  This is because when the peer reads a message on its socket, it won't know if it's a notification or a response to an RPC.  To fix this, the notification peer should be unable to make RPCs, and the handle_calls_from_peers code doesn't have to check for this case, or check for incoming connections on the socket, etc.  It just has to send messages.  Additionally since notifications are asynchronous, notification peers should probably have a queue of messages instead of just one message buffer.  WDYT?

wingo commented on src/program/alarms/listen/README:
```diff
@@ -0,0 +1,18 @@
+
+If the --socket argument is provided, it will accept a single client over the
+unix socket. If it has not been provided, it will accept commands over stdin
+and display them on stdout.
```
I think it would be OK to support multiple clients.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -527,6 +542,63 @@ end
+         local buf, len = head(peer.queue)
+         peer.pos = peer.pos or 0
+         local count, err = peer.fd:write(buf + peer.pos,
+                                          len - peer.pos)
```
You can actually do `ffi.cast('uint8_t*', queue[1]) + peer.pos` here to avoid the allocation in `head`.

dpino replied:
This is fixed although I changed `head` to return a pointer instead of doing an allocation, as suggested.

wingo commented on src/lib/yang/snabb-config-leader-v1.yang:
```diff
@@ -138,6 +138,16 @@ module snabb-config-leader-v1 {
 
+  rpc attach-notification-listener {
+    input {
+      leaf schema { type string; mandatory true; }
```
Are `schema` and `revision` actually used?

wingo reviewed (approved):
lgtm with some questions.  we should consider rewriting in terms of #1060 later.

dpino commented:
@wingo If it's ready to land I'd like to squash some commits before merging. Please, let me know if the last changes are OK.

dpino committed:
Push notifications over socket using a queue

Notifications are retrieved from alarms state and converted to text
messages.  Text messages are enqueued into each peer's notification
queue.  Later the peers queues is iterated and messages are transmitted
over each peer's socket.  Each peer keeps track of the latest position of
the last message transmitted.
